### ISSUE TITLE
Installing question

### ISSUE THREAD

**Issue by interose** [shamblett/mqtt_client#10]
Hi, I tried to install the package but it gives me the following error:
```
Running "flutter packages get" in myapp...
Package mqtt_client has no versions that match >=1.5.0 <2.0.0 derived from:
- myapp depends on version ^1.5.0
pub get failed (1)
```

I tested it with a fresh new app:
```
Running "flutter packages get" in myapp...           18,1s

[✓] Flutter is fully installed. (Channel beta, v0.1.5, on Mac OS X 10.13.3 17D102, locale de-DE)
[✗] Android toolchain - develop for Android devices is not installed.
[✓] iOS toolchain - develop for iOS devices is fully installed. (Xcode 9.2)
[✗] Android Studio is not installed. (not installed)
[✓] VS Code is fully installed. (version 1.21.1)
[✓] Connected devices is fully installed. (1 available)
```

I also tried to install version 1.4 or 1.3, but I always get the same message.
Contents of pubspec.yaml:
```
name: myapp
description: A new Flutter project.

dependencies:
  flutter:
    sdk: flutter

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^0.1.0
  mqtt_client: ^1.5.0

dev_dependencies:
  flutter_test:
    sdk: flutter
```

Any help or hint is greatly appreciated!

**Comment by shamblett** [comment]
Hmm, Iv'e never installed on Flutter but I guess that shouldn't matter, try relaxing the version number, what happens if you specify '<=2.0.0' ?

**Comment by interose** [comment]
Ah okay  - sorry. I searched via https://pub.dartlang.org/ and found your lib which is marked with the flutter badge.

I tried to add your lib via
```
dependencies:
  flutter:
    sdk: flutter

  # The following adds the Cupertino Icons font to your application.
  # Use with the CupertinoIcons class for iOS style icons.
  cupertino_icons: ^0.1.0
  mqtt_client: '<=2.0.0'
```
but It gives me the following
```
> flutter packages get
Running "flutter packages get" in myapp...
Package mqtt_client requires SDK version >=1.8.0 <2.0.0-dev.infinity but the current SDK is 2.0.0-edge.0d5cf900b021bf5c9fa593ffa12b15bcd1cc5fe0.
pub get failed (1)
```

Because flutter is really new and there is no stable branch nor I could choose a special SDK version, I guess this is to early ;)

No big deal, I already have a working mqtt swift example and tried to create a demo app with react native and/or flutter.

**Comment by krconv** [comment]
I relaxed the version number as suggested, which allows including it into a flutter project. https://github.com/krconv/mqtt_client/commit/4b9608ba13e4e8134f4ffe4d82bc4905ca1cad96

Unfortunately it looks like `eventables` has some dependency on `dart:mirrors`, which the flutter team has decided won't be supported: https://github.com/flutter/flutter/issues/1150

**Comment by krconv** [comment]
Looks like `eventable` could be easily modified to remove the need for `mirrors`; not sure if it is still being maintained, but it would be nice to be able to change it instead of `mqtt_client` because it is the only package I can find with the Emitter and Detector style.

See more: https://github.com/0xor1/eventable/issues/20

**Comment by shamblett** [comment]
Lets see if the issue you've raised above on eventables with the maintainergains any traction, as I said above the client doesn't need the Emitter/Detector style per se, any eventing package will do as long as it achieves the same overall goal however eventables is nice to use here, probably why I chose it.
I'll have a look at the alternatives.

There is another rather heavy handed way to proceed and that's to take the eventables code and create a new package without a mirrors dependency and push it up to pub. I did this with the json_object package, this relied on mirrors and its a package I use extensively in some of my other Dart projects, so I took the code, removed the mirrors dependency and published it as json_object_lite so others could use it also if needed. I wouldn't recommend this however as fix in all cases where we need to remove mirrors dependencies, it has to be maintained etc.

Another way to do it is take the eventables code, remove mirrors and just add the code to the client package, taking away the reliance on eventables, downside is more code to maintain in the client and we should really try and use the Dart ecosystem but it would remove the mirrors dependency. If it opens up the client to usage in Flutter it might be worth it.  I'll have a look at your eventables repo

**Comment by shamblett** [comment]
[event_bus](https://pub.dartlang.org/packages/event_bus),  may be an alternative, it doesn't use mixins but seems to achieve the same end result.

**Comment by shamblett** [comment]
I've created an new branch named eventbus where I've replaced eventable with event_bus, all the unit tests are running unchanged so it looks good, I need to run the client against some test brokers but this looks hopeful

**Comment by shamblett** [comment]
OK, I've tested this update with my IOT home project and it looks OK, I'll just prep it for release 1.6 then update pub

**Comment by shamblett** [comment]
OK, 1.6.0 pushed to pub, please give this try now.

**Comment by kylesnipes** [comment]
I'm completely new to Flutter and Dart. I'm trying to install this with the latest version of the Flutter SDK (Beta 2) and get the following error: 
"Package observable has no versions that match >=0.20.4+1 <0.21.0 derived from:
- mqtt_client 1.6.0 depends on version >=0.20.4+1 <0.21.0"

I'm not depending on observable directly. Any thoughts?

**Comment by shamblett** [comment]
mqtt_client depends on observable, the pub page for this package seems to show this version constraint should be valid, however its now not the latest version of this package. I'll re-publish the package with the latest observable version ASAP. In the mean time you could download mqtt_client from github and just update this locally to see if this solves the problem.

**Comment by shamblett** [comment]
I've updated qnd re-published mqtt_client to use the latest Observable, please pull this and see how you get on

**Comment by ravikumarpinjala** [comment]
import 'dart:async';
import 'dart:io';
import 'package:flutter/material.dart';
import 'package:mqtt_client/mqtt_client.dart' as MQTT;

uture<int> MQTTConnect() async {
  final MQTT.MqttClient client = new MQTT.MqttClient("127.0.1.1", "clientId");
  client.setProtocolV311();
  client.logging(true);
  

await client.connect();
  if (client.connectionState == MQTT.ConnectionState.connected) {
    print("iotcore client connected");
  } else {
    print(
        "ERROR iotcore client connection failed - disconnecting, state is ${client
            .connectionState}");
    client.disconnect();
  }
  return 0;
}

void main() => runApp(new MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    //DiscoverServer();
    Future<int> mqtt = MQTTConnect();
    mqtt.then((onValue) {
      print("MQTT Connected");
    }).catchError((onError){
      print("MQTT Chrash: $onError");
    });

    return new MaterialApp(
      title: 'Flutter Demo',
      theme: new ThemeData(
        // This is the theme of your application.
        //
        // Try running your application with "flutter run". You'll see the
        // application has a blue toolbar. Then, without quitting the app, try
        // changing the primarySwatch below to Colors.green and then invoke
        // "hot reload" (press "r" in the console where you ran "flutter run",
        // or press Run > Flutter Hot Reload in IntelliJ). Notice that the
        // counter didn't reset back to zero; the application is not restarted.
        primarySwatch: Colors.green,
      ),
      home: new MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  // This widget is the home page of your application. It is stateful, meaning
  // that it has a State object (defined below) that contains fields that affect
  // how it looks.

  // This class is the configuration for the state. It holds the values (in this
  // case the title) provided by the parent (in this case the App widget) and
  // used by the build method of the State. Fields in a Widget subclass are
  // always marked "final".

  final String title;

  @override
  _MyHomePageState createState() => new _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;
  
  void _incrementCounter() {
    setState(() {
      // This call to setState tells the Flutter framework that something has
      // changed in this State, which causes it to rerun the build method below
      // so that the display can reflect the updated values. If we changed
      // _counter without calling setState(), then the build method would not be
      // called again, and so nothing would appear to happen.
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    // This method is rerun every time setState is called, for instance as done
    // by the _incrementCounter method above.
    //
    // The Flutter framework has been optimized to make rerunning build methods
    // fast, so that you can just rebuild anything that needs updating rather
    // than having to individually change instances of widgets.
    return new Scaffold(
      appBar: new AppBar(
        // Here we take the value from the MyHomePage object that was created by
        // the App.build method, and use it to set our appbar title.
        title: new Text(widget.title),
      ),
      body: new Center(
        // Center is a layout widget. It takes a single child and positions it
        // in the middle of the parent.
        child: new Column(
          // Column is also layout widget. It takes a list of children and
          // arranges them vertically. By default, it sizes itself to fit its
          // children horizontally, and tries to be as tall as its parent.
          //
          // Invoke "debug paint" (press "p" in the console where you ran
          // "flutter run", or select "Toggle Debug Paint" from the Flutter tool
          // window in IntelliJ) to see the wireframe for each widget.
          //
          // Column has various properties to control how it sizes itself and
          // how it positions its children. Here we use mainAxisAlignment to
          // center the children vertically; the main axis here is the vertical
          // axis because Columns are vertical (the cross axis would be
          // horizontal).
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            new Text(
              'You have pushed the button this many times:',
            ),
            new Text(
              '$_counter',
              style: Theme.of(context).textTheme.display1,
            ),
          ],
        ),
      ),
      floatingActionButton: new FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: new Icon(Icons.add),
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}

ConnectionState is conflicting with async.dat function.
When I run "flutter run" the below error is coming.
MQTT Chrash: type '(MessageReceived) => void' is not a subtype of type '(dynamic) => void'

**Comment by shamblett** [comment]
Yes, the pub page for the package is also pulling this :

Strong-mode analysis of lib/src/mqtt_client_subscriptions_manager.dart gave the following warning:

line: 44 col: 47
A function of type '(MessageReceived) → void' can't be assigned to a location of type '(dynamic) → void'.

Unfortunately the 'strong mode' it refers to is Dart 2.0 strong mode, mqtt_client is currently for Dart 1.0(as are all my packages) the reason being Dart 2.0 is not yet released as a production build. When this happens I'll update all my packages to suit. If I get time I'll download a dev version of Dart 2.0 and see if I can fix these but I'm loath to do this until dart 2.0 is released.

**Comment by cwitte4191** [comment]
Steve, 

I'm getting the following error trying to consume mqtt_client 1.7.0 (which i believe to be current).  It seems related to this thread.   I was getting a less verbose version of the error, and then ran flutter upgrade.  Now i get a more detailed message but it still fails.   I can make the error go away if I use a local copy of mqtt_client (forked from master) and change pubspec.yaml to use "observable: any".  I'm not sure where the reference to mqtt_client version 1.6.1 is from.  (my pubspec.yaml specifies ^1.7.0



The error is:
`Running "flutter packages get" in flutter0322...
Because every version of flutter_test from sdk depends on quiver 0.29.0+1 and observable >=0.21.0
  depends on quiver >=0.24.0 <0.29.0, flutter_test from sdk is incompatible with observable
  >=0.21.0.
And because mqtt_client >=1.6.1 depends on observable >=0.21.0 <=1.0.0, flutter_test from sdk is
  incompatible with mqtt_client >=1.6.1.
So, because flutter0322 depends on both mqtt_client ^1.7.0 and flutter_test any from sdk, version
  solving failed.
pub get failed (1)`


Thanks,
Chris.

**Comment by shamblett** [comment]
Ok, I've updated the pubspec to use "any" for Observable and republished the package at 1.7.1, sse how you get on.

The Flutter package resolver seems to be quite finnicky compared to the Dart one.

**Comment by cwitte4191** [comment]
Changing the dependency to "any" resolves the compile time issues, but now produces some run time errors that are related to the Observables (a subset of the error follows).  This probably would have happened with my local change, but I hadn't run that earlier (I had wanted to query your thoughts).
`Resolving dependencies...
Running 'gradlew assembleDebug'...
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client.dart:143:3: Error: Type 'ChangeNotifier' not found.
compiler message:   ChangeNotifier<MqttReceivedMessage> listenTo(String topic, MqttQos qosLevel) {
compiler message:   ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscription.dart:25:3: Error: Type 'ChangeNotifier' not found.
compiler message:   ChangeNotifier<MqttReceivedMessage> observable;
compiler message:   ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscriptions_manager.dart:29:15: Error: Type 'ChangeNotifier' not found.
compiler message:   Map<String, ChangeNotifier<MqttReceivedMessage>> messagesReceived =
compiler message:               ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscriptions_manager.dart:48:3: Error: Type 'ChangeNotifier' not found.
compiler message:   ChangeNotifier<MqttReceivedMessage> registerSubscription(String topic,
compiler message:   ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscriptions_manager.dart:58:3: Error: Type 'ChangeNotifier' not found.
compiler message:   ChangeNotifier<MqttReceivedMessage> tryGetExistingSubscription(String topic) {
compiler message:   ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscriptions_manager.dart:72:3: Error: Type 'ChangeNotifier' not found.
compiler message:   ChangeNotifier<MqttReceivedMessage> createNewSubscription(String topic,
compiler message:   ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscriptions_manager.dart:114:3: Error: Type 'ChangeNotifier' not found.
compiler message:   ChangeNotifier<MqttReceivedMessage> createObservableForSubscription(
compiler message:   ^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscription.dart:25:3: Error: 'ChangeNotifier' isn't a type.
compiler message:   ChangeNotifier<MqttReceivedMessage> observable;
compiler message:   ^^^^^^^^^^^^^^
compiler message: file:///Users/cwitte/.pub-cache/hosted/pub.dartlang.org/mqtt_client-1.7.1/lib/src/mqtt_client_subscriptions_manager.dart:29:15: Error: 'ChangeNotifier' isn't a type.
compiler message:   Map<String, ChangeNotifier<MqttReceivedMessage>> messagesReceived =`

**Comment by shamblett** [comment]
Well its saying it can't find the ChangeNotifier class from Observable as you've probably guessed, although why this should be I don't know., its as though the Observables package hasn't been compiled or something, maybe the Flutter guys could help.

**Comment by shamblett** [comment]
Sounds as though the MqttConnection hasn't been constructed properly, disconnectRequested is only used on disconnection from the MQTT server. When you say 'running the test alone is fine' do you mean its OK in the Dart VM? This looks like a flutter specific thing, have you tried debugging it?

**Comment by cwitte4191** [comment]
I'm still having issues with mqtt_client and ChangeNotifier.  I now suspect that the issue is a namespace collision with the ChangeNotifier defined in the Observable package and the ChangeNotifier implementation in flutter.   Is there a way to qualify the ChangeNotifier references in mqtt_client so that the Observable implementation is specified?

The relevant (and non compatible) change_notifier implementations are:

https://github.com/dart-lang/observable/blob/master/lib/src/change_notifier.dart
https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/foundation/change_notifier.dart

**Comment by shamblett** [comment]
Yep, I've prefixed the Observable library in the client, this should make it clear which change notifier we are using, version 1.7.2 published.

**Comment by shamblett** [comment]
Fixed in latest client(3.2.1), this now works OK with flutter
